Add flag for private cloud support

Afterglow opened the pull request:
This is a potential fix for #3 maintains backwards compatibility and provides a new private flag when creating a client.

adamleff commented:
@Afterglow - thanks for your PR!

A few things:

Have you or your company signed the [CLA](https://supermarket.chef.io/become-a-contributor)?  If so, can you provide your supermarket ID?  If you need to send it in private, you can email me directly (adamleff at chef dot io)

Now, structurally and functionally, this is great! I'd like to suggest a few changes:

 * Since `private` is a bit of a loaded term, I'd like to request we change the option as well as the instance variable to `@private_cloud` so it's 100% clear what it's for.
 * Rather than using the instance variable, I'd like to recommend we create a helper method called `private_cloud?` that returns the contents of `@private_cloud`.  This will allow you to mock the method in your tests and avoid needing to set up a whole separate client object for testing (as done at https://github.com/chef-partners/oracle-cloud-client/pull/4/files#diff-b00e26cc02e5f5b6f69d826564a6f280R54). 

Catch me at the summit or online and let's pair!

gstankovski commented:
Hi guys,

Thanks for putting this together. I want to ask whether it should be more specific with the naming of this.  This client is for IaaS services provided by Oracle Public and Private Cloud.

There are many other Cloud services Oracle provides, and it makes sense that we start putting together support for these. To name a few (Oracle Java Cloud Service, Oracle DB Cloud Service, Oracle Integration Cloud Service, etc ...).  These all have similar REST interfaces, and it would be great if we can pull together clients to support all of these.  Not sure if they go into one big client package or have individual ones for each cloud service Oracle provides.

@adamleff - FYI - we've also been doing some work with Sanjay Sinha and Ajay Seetheram at Oracle (your name was mentioned)

Cheers,
Goran

Afterglow commented:
@adamleff i've updated my branch, see what you think i've added a private_cloud? method which i've used instead of the private variable and i've removed the duplicate client from the tests

Afterglow commented:
oh and my supermarket id is afterglow I should have a CLA signed on it

adamleff commented:
@gstankovski I appreciate your concerns regarding the naming of the gem.  However, I'm going to leave the name as-is.  It was specifically created for the product marketed as "Oracle Cloud" found at "cloud.oracle.com"  I'm happy to include some additional specificity to the gem description and README if we find it's getting confusing.  But since we have active users of the already-named gem as well as projects depending on it by name, renaming will cause more hassle than it's worth.

adamleff commented:
CLA for @Afterglow is confirmed.

adamleff commented:
@Afterglow this looks good to me!  :+1: 

@chef-partners/partner-engineering - another +1 on this please.

gstankovski commented:
@adamleff Ok. Was just pointing out that Oracle Cloud is an umbrella product offering that contains many PaaS, SaaS, and now IaaS services.  The client in question is for the IaaS offering.  Its analogous to creating a client called AWS that only provides EC2 capabilities, even though there are many more services available.

I'd like to start putting together GEM clients for some of the other PaaS services available in Oracle Cloud, namely, the Oracle Java Cloud Service (JCS), Oracle Database Cloud Services (DBaaS) and Oracle SOA Cloud Service (SOA), Oracle Integration Cloud Service (ICS).  They will all have similar REST API interfaces.

What shall the naming of these GEMs be?

oraclecloud-jcs
oraclecloud-dbaas
oraclecloud-soa
oraclecloud-ics

Are these going to be acceptable? I think they may sense.

adamleff commented:
@gstankovski I certainly understand your point. However, this client gem exists solely for the purpose of supporting the Chef knife and test kitchen plugins that we authored so we do not have to maintain this code in multiple places. If others wish to use it, that makes me really happy! However, we will not be renaming the gem. 

Your proposed names make sense to me.

mccoold commented:
Gowan,

I have a gem for the pass api's called OPC. I have also started to tie it in to knife with a knife plugin as well, the integration into knife is just starting.  

Thank You,

Daryn


> On Nov 8, 2015, at 6:03 PM, Goran Stankovski <notifications@github.com> wrote:
> 
> @adamleff Ok. Was just pointing out that Oracle Cloud is an umbrella product offering that contains many PaaS, SaaS, and now IaaS services. The client in question is for the IaaS offering. Its analogous to creating a client called AWS that only provides EC2 capabilities, even though there are many more services available.
> 
> I'd like to start putting together GEM clients for some of the other PaaS services available in Oracle Cloud, namely, the Oracle Java Cloud Service (JCS), Oracle Database Cloud Services (DBaaS) and Oracle SOA Cloud Service (SOA), Oracle Integration Cloud Service (ICS). They will all have similar REST API interfaces.
> 
> What shall the naming of these GEMs be?
> 
> oraclecloud-jcs
> oraclecloud-dbaas
> oraclecloud-soa
> oraclecloud-ics
> 
> Are these going to be acceptable? I think they may sense.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mccoold commented:
Sorry,  the  gem is for the PaaS services, spell check got in the way there.  JCS, DBCS, and object storage functionality is all available there, SOA is coming soon, Api's not fully published yet.  I am also adding document cloud service to the gem now. I typically add the functionality to the command line gem at the same time followed by the knife plugin, which is very basic at this point. I am reviewing the code here to see how easy it would be to integrate into this base.   I am a newbie compared to the experts at chef 

Thank You,

Daryn


> On Nov 9, 2015, at 1:36 PM, Daryn McCool <mdaryn@hotmail.com> wrote:
> 
> Gowan,
> 
> I have a gem for the pass api's called OPC. I have also started to tie it in to knife with a knife plugin as well, the integration into knife is just starting.  
> 
> Thank You,
> 
> Daryn
> 
> 
>> On Nov 8, 2015, at 6:03 PM, Goran Stankovski <notifications@github.com> wrote:
>> 
>> @adamleff Ok. Was just pointing out that Oracle Cloud is an umbrella product offering that contains many PaaS, SaaS, and now IaaS services. The client in question is for the IaaS offering. Its analogous to creating a client called AWS that only provides EC2 capabilities, even though there are many more services available.
>> 
>> I'd like to start putting together GEM clients for some of the other PaaS services available in Oracle Cloud, namely, the Oracle Java Cloud Service (JCS), Oracle Database Cloud Services (DBaaS) and Oracle SOA Cloud Service (SOA), Oracle Integration Cloud Service (ICS). They will all have similar REST API interfaces.
>> 
>> What shall the naming of these GEMs be?
>> 
>> oraclecloud-jcs
>> oraclecloud-dbaas
>> oraclecloud-soa
>> oraclecloud-ics
>> 
>> Are these going to be acceptable? I think they may sense.
>> 
>> —
>> Reply to this email directly or view it on GitHub.
>>

larryebaum commented:
Reviewed per request from @adamleff , :+1:

adamleff commented:
@Afterglow version 1.1.0 of the gem, which includes this PR, is available on RubyGems.

Afterglow commented:
thanks @adamleff
